permute/unpermute kernel for moe optimization

CalebDu opened the pull request:
`moe_permute` kernel expands and oreders token in activation to gather uncontinuous tokens for each expert. And then call grouped-gemm for moe speedup.
`moe_unpermute` kernel reduces  expanded grouped-gemm output and scales with `topk_weight`.
<img width="646" alt="image" src="https://github.com/user-attachments/assets/4b6b7691-8772-474a-b803-ad3a34c6417b" />

This implementation refers to  moe kernel in  tensorrt-llm in archive https://github.com/BBuf/tensorrt-llm-moe/tree/master.
Currently, unsupport  Expert-Parallelism with `expert_map`, will follow up with updates.

CalebDu committed:
implement moe permute kernel

Signed-off-by: Caleb_Du <Caleb_Du@zju.edu.cn>

CalebDu committed:
implement moe unpermute

Signed-off-by: Caleb_Du <Caleb_Du@zju.edu.cn>

CalebDu committed:
update code and fix pre-commit

Signed-off-by: Caleb_Du <Caleb_Du@zju.edu.cn>

gzy19990617 commented:
请问这个算子是为后面做group_gemm做准备工作嘛

CalebDu commented:
> 请问这个算子是为后面做group_gemm做准备工作嘛
yes

gzy19990617 commented:
请问预计什么时候会支持呐

---- Replied Message ----
| From | ***@***.***> |
| Date | 03/12/2025 16:54 |
| To | vllm-project/vllm ***@***.***> |
| Cc | gaoziyuan ***@***.***>,
Comment ***@***.***> |
| Subject | Re: [vllm-project/vllm] permute/unpermute kernel for moe optimization (PR #14568) |

请问这个算子是为后面做group_gemm做准备工作嘛

yes

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you commented.Message ID: ***@***.***>

CalebDu left a comment (vllm-project/vllm#14568)

请问这个算子是为后面做group_gemm做准备工作嘛

yes

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you commented.Message ID: ***@***.***>

CalebDu commented:
> 请问预计什么时候会支持呐
> […](#)
> ---- Replied Message ---- | From | ***@***.***> | | Date | 03/12/2025 16:54 | | To | vllm-project/vllm ***@***.***> | | Cc | gaoziyuan ***@***.***>, Comment ***@***.***> | | Subject | Re: [vllm-project/vllm] permute/unpermute kernel for moe optimization (PR #14568) | 请问这个算子是为后面做group_gemm做准备工作嘛 yes — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you commented.Message ID: ***@***.***> CalebDu left a comment ([vllm-project/vllm#14568](https://github.com/vllm-project/vllm/pull/14568)) 请问这个算子是为后面做group_gemm做准备工作嘛 yes — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you commented.Message ID: ***@***.***>

我也不清楚具体release的时间

CalebDu committed:
add expert_map support for moe permute

Signed-off-by: Caleb_Du <Caleb_Du@zju.edu.cn>

CalebDu committed:
add expert_map support for moe unpermute

Signed-off-by: Caleb_Du <Caleb_Du@zju.edu.cn>

CalebDu commented:
update `expert_map` support for expert parallelism.

CalebDu commented:
This PR submits 2 kernel `moe_permute` and `moe_unpermute`.
`moe_permute` includes:
- `preprocess_topk_id_launcher`(optional for EP)：`preprocess_topk_id` is to map global expert id to local expert id for each ep rank. For example, 4 expert with ep_size=2. ep_rank=1 owns global expert id [2,3] with `expert_map=[-1, -1, 0, 1]`, `preprocess_topk_id`  process `topk_ids` and map global expert id [2, 3] to local_expert id [0, 1] and map global expert id [0, 1] ( not in ep rank=1)  to [4, 5] by plus `n_expert`. This map operation is to make local expert high priority in following sort `topk_ids`  and scan `expert_first_token_offset` for local expert. So sorting `topk_ids` can get results in  global expert id [2, 3, 0, 1] (local expert id[0, 1,4,5] ) order and scanning get `expert_first_token_offset`  of  local expert id 0 1(global expert id 2 3).
- `sortAndScanExpert`：call `cub key-value radix-sort ` to sort local expert_id in `topk_id` and get mapping idx for expanding and permuting input activation. And call `computeExpertFirstTokenOffset` to get first token offset in permuted activation m-dim for following group gemm.
- `expandInputRowsKernelLauncher` : expand and permute input activation with idx map from `sortAndScanExpert`. And get mapping idx for `moe_unpermute`.


`moe_unpermute` includes:
- `finalizeMoeRoutingKernelLauncher`：reduce permuted activation  with `topk_idx` and sort by original token order.

bnellnm commented:
Hi @CalebDu , thanks for working on this.  I think this might help me out with https://github.com/vllm-project/vllm/pull/13932.  I am currently trying to use your PR to do the permute/unpermute steps needed for the DeepGemm grouped gemm kernel.

I did run into a test failure with certain problem sizes that are not in the original test, e.g.
```
@pytest.mark.parametrize("n_token", [2048])
@pytest.mark.parametrize("n_hidden", [7168])
@pytest.mark.parametrize("n_expert", [64])
@pytest.mark.parametrize("topk", [6])
@pytest.mark.parametrize("dtype", [torch.bfloat16])
@pytest.mark.parametrize("ep_size", [1])
def test_moe_permute_unpermute(
    n_token: int, n_hidden: int, topk: int, n_expert: int, ep_size: int, dtype: torch.dtype
)

========================================================= FAILURES ==========================================================
____________________________________ test_moe_permute_unpermute[1-dtype0-6-64-7168-2048] ____________________________________

n_token = 2048, n_hidden = 7168, topk = 6, n_expert = 64, ep_size = 1, dtype = torch.bfloat16

    @pytest.mark.parametrize("n_token", [2048])
    @pytest.mark.parametrize("n_hidden", [7168])
    @pytest.mark.parametrize("n_expert", [64])
    @pytest.mark.parametrize("topk", [6])
    #@pytest.mark.parametrize("dtype", [torch.float16, torch.bfloat16])
    @pytest.mark.parametrize("dtype", [torch.bfloat16])
    #@pytest.mark.parametrize("ep_size", EP_SIZE)
    @pytest.mark.parametrize("ep_size", [1])
    def test_moe_permute_unpermute(
        n_token: int, n_hidden: int, topk: int, n_expert: int, ep_size: int, dtype: torch.dtype
    ):
        ep_rank= 0 #np.random.randint(0, ep_size)
        expert_map = None
        n_local_expert = n_expert
        if(ep_size != 1) :
            n_local_expert, expert_map = determine_expert_map(ep_size, ep_rank, n_expert)
            expert_map = expert_map.cuda()
        start_expert = n_local_expert * ep_rank
        current_platform.seed_everything(0)
        hidden_states = torch.randn((n_token, n_hidden), device="cuda").to(dtype)
        gating_output = torch.randn((n_token, n_expert), device="cuda").to(dtype)
        topk_weights, topk_ids, token_expert_indices = fused_topk(
            hidden_states, gating_output, topk, False
        )

        gold0, gold1, gold2 = torch_permute(hidden_states,
                                            topk_ids,
                                            token_expert_indices,
                                            topk, n_expert,n_local_expert,
                                            start_expert, expert_map=expert_map)

        result0, result1, result2 = moe_permute(hidden_states,
                                                topk_weights, topk_ids,
                                                token_expert_indices,
                                                topk, n_expert, n_local_expert, expert_map
                                                )
        # print(gold0, result0)
        #print(gold1, result1)
        # print(gold2, result2)

        # check expert_first_token_offset
>       torch.testing.assert_close(gold1,
                                   result1,
                                   atol=0,
                                   rtol=0)
E       AssertionError: Tensor-likes are not equal!
E       
E       Mismatched elements: 2 / 65 (3.1%)
E       Greatest absolute difference: 106 at index (3,)
E       Greatest relative difference: 0.150141641497612 at index (3,)

tests/kernels/test_moe_permute_unpermute.py:129: AssertionError
================================================== short test summary info ==================================================
FAILED tests/kernels/test_moe_permute_unpermute.py::test_moe_permute_unpermute[1-dtype0-6-64-7168-2048] - AssertionError: Tensor-likes are not equal!
===================================================== 1 failed in 3.08s =====================================================
```

Do you have any idea where the problem might be?

CalebDu committed:
fix mismatch and add more test case

Signed-off-by: Caleb_Du <Caleb_Du@zju.edu.cn>

CalebDu commented:
@bnellnm  this bug is caused by the workspace is too small for cub radix_sort. After I expand workspace, I fix it.

bnellnm commented:
> @bnellnm this bug is caused by the workspace is too small for cub radix_sort. After I expand workspace, I fix it.

Thanks!

CalebDu committed:
support  align_block_size for contiguous group gemm in deepgemm

Signed-off-by: Caleb_Du <Caleb_Du@zju.edu.cn>

CalebDu commented:
Add `align_block_size` support for contiguous group gemm in deepgemm.
Round up token amount in each expert to `align_block_size`  and scan aligned amount  to get `align_first_token_offset` for each expert.  So in permuted_ hidden , `align_permuted_row_idx ` = `permuted_row_idx` - `first_token_offset`+ `align_first_token_offset`.

bnellnm commented:
@CalebDu , thanks for adding the blocking support!  I've been working on integrating the new version with DeepGemm but I'm running into problems with the `m_indices`.  I don't think they are being computed correctly.  I think the best way to test for correctness is to use `moe_align_block_size` like this:
```
block_m = 128
_, expert_ids, _ = moe_align_block_size(curr_topk_ids, block_m, global_num_experts, expert_map)
m_indices = torch.repeat_interleave(expert_ids, block_m, dim=0)
```

CalebDu commented:
> @CalebDu , thanks for adding the blocking support! I've been working on integrating the new version with DeepGemm but I'm running into problems with the `m_indices`. I don't think they are being computed correctly. I think the best way to test for correctness is to use `moe_align_block_size` like this:
> 
> ```
> block_m = 128
> _, expert_ids, _ = moe_align_block_size(curr_topk_ids, block_m, global_num_experts, expert_map)
> m_indices = torch.repeat_interleave(expert_ids, block_m, dim=0)
> ```

I notice deepgemm update document about `m_grouped_gemm_fp8_fp8_bf16_nt_contiguous`  in commit https://github.com/deepseek-ai/DeepGEMM/commit/ded740f736570c000b7063d7774f7c18b5d1ecd6. In original version, `m_indices` mark `-1 `  for aligned invalid row ans skip it. So I fill `-1` for all padding row in `m_indices`. It seems be not same as `moe_align_block_size`. I will update this part to fill `m_indices` with `expert_id` for all row rather than only valid row.
I have a question, deepgemm will check `m_indices.numel()` == `lhs.shape[0]`, but `sorted_ids.numel()`!=`m_indices.numel()` in following code. 
```
block_m = 128
sorted_ids, expert_ids, _ = moe_align_block_size(curr_topk_ids, block_m, global_num_experts, expert_map)
m_indices = torch.repeat_interleave(expert_ids, block_m, dim=0)
```

bnellnm commented:
> > @CalebDu , thanks for adding the blocking support! I've been working on integrating the new version with DeepGemm but I'm running into problems with the `m_indices`. I don't think they are being computed correctly. I think the best way to test for correctness is to use `moe_align_block_size` like this:
> > ```
> > block_m = 128
> > _, expert_ids, _ = moe_align_block_size(curr_topk_ids, block_m, global_num_experts, expert_map)
> > m_indices = torch.repeat_interleave(expert_ids, block_m, dim=0)
> > ```
> 
> I notice deepgemm update document about `m_grouped_gemm_fp8_fp8_bf16_nt_contiguous` in commit [deepseek-ai/DeepGEMM@ded740f](https://github.com/deepseek-ai/DeepGEMM/commit/ded740f736570c000b7063d7774f7c18b5d1ecd6). In original version, `m_indices` mark `-1 ` for aligned invalid row ans skip it. So I fill `-1` for all padding row in `m_indices`. It seems be not same as `moe_align_block_size`. I will update this part to fill `m_indices` with `expert_id` for all row rather than only valid row. I have a question, deepgemm will check `m_indices.numel()` == `lhs.shape[0]`, but `sorted_ids.numel()`=`max_num_tokens_padded`!=`m_indices.numel()`= `round_up(max_num_tokens_padded, align_block_size )` in following code.
> 
> ```
> block_m = 128
> sorted_ids, expert_ids, _ = moe_align_block_size(curr_topk_ids, block_m, global_num_experts, expert_map)
> m_indices = torch.repeat_interleave(expert_ids, block_m, dim=0)
> ```
> 
> And `moe_align_block_size` will also fill `m_indices` with `-1` for the expert that is invalid in `ep_map`. So we'd better to `view/slice` only aligned row in `permuted_hidden`and `m_indices` for valid expert to compute grouped gemm. Like following code.
> 
> ```
> permuted_hidden = permuted_hidden[: align_first_token_offset[-1],...]
> m_indices = m_indices[: align_first_token_offset[-1]]
> ```
> 
> Do you have a better idea?

The `-1` in the original DeepGemm documentation was a bug.  They never actually supported it (I ran into this while doing the initial integration).  In the non expert_map case, `moe_align_block_size` fills in unused sections with 0s.  I haven't really tested DeepGemm with an explicit expert_map but I guess it would be ok to replace any -1 with 0 (or any other valid expert index) since it will get dropped at the end anyway.

The above code snippet works for DeepGemm (without an explicit expert_map).  I'll have to add some tests + code for the user suppled expert_map.  At the moment unused values get chopped off at the end by the inverse map.  But it would probably be better if we could chop them off up front like you suggest.  I'm not sure of a clean/simple way to do that since they'd need to be chopped off in 128 element sized chunks (at least for DeepGemm).

CalebDu committed:
for each local valid expert,fill padding row with expert_id in `m_indices`rather than -1,

Signed-off-by: Caleb_Du <Caleb_Du@zju.edu.cn>

CalebDu commented:
> The `-1` in the original DeepGemm documentation was a bug. They never actually supported it (I ran into this while doing the initial integration). In the non expert_map case, `moe_align_block_size` fills in unused sections with 0s. I haven't really tested DeepGemm with an explicit expert_map but I guess it would be ok to replace any -1 with 0 (or any other valid expert index) since it will get dropped at the end anyway.
> 
> The above code snippet works for DeepGemm (without an explicit expert_map). I'll have to add some tests + code for the user suppled expert_map. At the moment unused values get chopped off at the end by the inverse map. But it would probably be better if we could chop them off up front like you suggest. I'm not sure of a clean/simple way to do that since they'd need to be chopped off in 128 element sized chunks (at least for DeepGemm).

I update code about fill padding row with `expert_id` in `m_indices` rather than -1 for each local  valid expert. I think filling 0 (or other valid expert index) in  `m_indices`  for non-local expert  is not a good idea. Because it may make a alot redundant compute and memory traffic for many tailing padding row in DeepGemm. 
If pass `align_block_size` to `moe_permute` kernel,  output `expert_first_token_offset` will align up to  `align_block_size`. For example，local expert [0,1] has [127, 129] tokens respectively. `expert_first_token_offset` will be` [0, 128, 384]` rather than `[0, 127, 256]`. You can use `expert_first_token_offset[-1]` directly to slice front valid row in `permuted_hidden` to skip -1 in `m_indices`.

bnellnm commented:
> > The `-1` in the original DeepGemm documentation was a bug. They never actually supported it (I ran into this while doing the initial integration). In the non expert_map case, `moe_align_block_size` fills in unused sections with 0s. I haven't really tested DeepGemm with an explicit expert_map but I guess it would be ok to replace any -1 with 0 (or any other valid expert index) since it will get dropped at the end anyway.
> > The above code snippet works for DeepGemm (without an explicit expert_map). I'll have to add some tests + code for the user suppled expert_map. At the moment unused values get chopped off at the end by the inverse map. But it would probably be better if we could chop them off up front like you suggest. I'm not sure of a clean/simple way to do that since they'd need to be chopped off in 128 element sized chunks (at least for DeepGemm).
> 
> I update code about fill padding row with `expert_id` in `m_indices` rather than -1 for each local valid expert and still fill -1 for non-local expert. I think filling 0 (or other valid expert index) in `m_indices` for non-local expert is not a good idea. Because it may make a lot redundant compute and memory traffic for many tailing padding row in DeepGemm. If pass `align_block_size` to `moe_permute` kernel, output `expert_first_token_offset` will align up to `align_block_size`. For example，local expert [0,1] has [127, 129] tokens respectively. `expert_first_token_offset` will be` [0, 128, 384]` rather than `[0, 127, 256]`. You can use `expert_first_token_offset[-1]` directly to slice front valid row in `permuted_hidden` to skip -1 in `m_indices`.

Cool, I didn't realize you could use the first offset tokens to exclude the unused bits in that case.  If that's the case I don't think it matters what goes into m_indices since they will be sliced off anyway.

CalebDu commented:
> Cool, I didn't realize you could use the first offset tokens to exclude the unused bits. If that's the case I don't think it matters what goes into m_indices since they will be sliced off anyway.

I can try it.

CalebDu committed:
add fill_invalid_expert to workaround deepgemm unsupport -1 in m_indices

Signed-off-by: Caleb_Du <Caleb_Du@zju.edu.cn>

bnellnm commented on CMakeLists.txt:
```diff
@@ -15,7 +15,7 @@ project(vllm_extensions LANGUAGES CXX)
 # CUDA by default, can be overridden by using -DVLLM_TARGET_DEVICE=... (used by setup.py)
 set(VLLM_TARGET_DEVICE "cuda" CACHE STRING "Target device backend for vLLM")
-
+set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
```
What does this flag do?

CalebDu replied:
This flag is to export compile_commands.json in build path for clangd language server parsing c++ code. So clangd can support Code hinting and code completion in vscode. This line is an auxiliary code that can be removed.

bnellnm commented on CMakeLists.txt:
```diff
@@ -602,6 +602,25 @@ if(VLLM_GPU_LANG STREQUAL "CUDA")
+  cuda_archs_loose_intersection(MOE_PERMUTE_ARCHS "8.0;8.6;8.7;8.9;9.0;10.0;10.1;12.0" "${CUDA_ARCHS}")
+  if (MOE_PERMUTE_ARCHS)
+    set(MARLIN_PERMUTE_SRC
+        "csrc/moe/permute_unpermute_kernels/dispatch.h" 
+        "csrc/moe/permute_unpermute_kernels/moe_permute_unpermute_kernel.h"
```
I don't think it's necessary to add header files.

bnellnm commented on CMakeLists.txt:
```diff
@@ -602,6 +602,25 @@ if(VLLM_GPU_LANG STREQUAL "CUDA")
 endif()
 
+if(VLLM_GPU_LANG STREQUAL "CUDA")
+  cuda_archs_loose_intersection(MOE_PERMUTE_ARCHS "8.0;8.6;8.7;8.9;9.0;10.0;10.1;12.0" "${CUDA_ARCHS}")
+  if (MOE_PERMUTE_ARCHS)
```
I think it might be cleaner and more forward compatible if the CUDA arch check was moved to C++.  We could skip registering the new ops for incompatible versions.

CalebDu replied:
Do You mean that use macro to skip register op like `#if __CUDA_ARCH__ > 800`?

bnellnm replied:
Yeah (also use it to skip compiling the implementation).  I think that will be more robust than checking version strings in cmake.

bnellnm commented on csrc/moe/moe_permute_unpermute_op.cu:
```diff
@@ -0,0 +1,139 @@
+    preprocessTopkIdLauncher(get_ptr<int>(topk_ids), n_token * topk,
+                             expert_map_ptr, n_expert, stream);
+  }
+  // std::cout << "tops id " << topk_ids << std::endl;
```
nit: remove debugging code, ditto elsewhere.

bnellnm commented on csrc/moe/torch_bindings.cpp:
```diff
@@ -51,6 +51,20 @@ TORCH_LIBRARY_EXPAND(TORCH_EXTENSION_NAME, m) {
       "int moe_block_size, bool replicate_input, bool apply_weights)"
       " -> Tensor");
+
+  m.def(
+      "moe_permute(Tensor! input, Tensor! topk_weight, Tensor! topk_ids,"
+      "Tensor! token_expert_indicies, Tensor? expert_map, int n_expert,"
+      "int n_local_expert,"
+      "int topk, int? align_block_size,Tensor! permuted_input, Tensor! "
+      "expert_first_token_offset, Tensor! src_row_id2dst_row_id_map, Tensor! "
+      "m_indices)->()");
+
+  m.def(
+      "moe_unpermute(Tensor! permuted_hidden_states, Tensor! topk_weights,"
+      "Tensor! topk_ids,Tensor! src_row_id2dst_row_id_map, Tensor! "
+      "expert_first_token_offset, int n_expert, int n_local_expert,int "
+      "topk, Tensor! hidden_states)->()");
```
We shouldn't mark all the inputs as mutating (`!`), only where necessary.

bnellnm commented on vllm/model_executor/layers/fused_moe/fused_moe.py:
```diff
@@ -1089,11 +1089,11 @@ def fused_topk(
 
-    del token_expert_indicies  # Not used. Will be used in the future.
-    return topk_weights, topk_ids
+    # del token_expert_indices  # Not used. Will be used in the future.
```
Might as well delete this line now.

bnellnm commented on vllm/model_executor/layers/fused_moe/moe_permute_unpermute.py:
```diff
@@ -0,0 +1,119 @@
+    expert_map: Optional[torch.Tensor] = None,
+    align_block_size:Optional[int]=None,
+    fill_invalid_expert:int = -1
+) -> list[torch.Tensor]:
```
Is there any reason not to make the return type an explicit tuple of tensors?  So users will know exactly how many outputs to expect?

bnellnm commented on vllm/model_executor/layers/fused_moe/moe_permute_unpermute.py:
```diff
@@ -0,0 +1,119 @@
+        dtype=hidden_states.dtype,
+        device=hidden_states.device,
+    )
+    m_indices = torch.empty(permuted_row_size,
+                            dtype=torch.int32,
+                            device=hidden_states.device,).fill_(fill_invalid_expert)
```
I think you can do `torch.full(permuted_row_size, fill_invalid_expert, ...)`.

bnellnm commented on csrc/moe/permute_unpermute_kernels/moe_permute_unpermute_kernel.h:
Can you move all the non-trivial definitions from this file into a .cu file?

bnellnm commented on csrc/moe/permute_unpermute_kernels/moe_permute_unpermute_kernel.h:
```diff
@@ -0,0 +1,508 @@
+__device__ inline int64_t findTotalEltsLessThanTarget(T const* sorted_indices,
+                                                      int64_t const arr_length,
+                                                      T const target) {
+  int64_t low = 0, high = arr_length - 1, target_location = -1;
+  while (low <= high) {
+    int64_t mid = (low + high) / 2;
+
+    if (sorted_indices[mid] >= target) {
+      high = mid - 1;
+    } else {
+      low = mid + 1;
+      target_location = mid;
+    }
+  }
+  return target_location + 1;
```
could you use `std::binary_search` here instead?

bnellnm commented on csrc/moe/moe_permute_unpermute_op.cu:
```diff
@@ -0,0 +1,139 @@
+  // and scan local expert_first_token_offset for each ep rank for next group
+  // gemm.
+  if (expert_map.has_value()) {
+    int* expert_map_ptr = reinterpret_cast<int*>(expert_map.value().data_ptr());
```
Can you use `get_ptr<int>(expert_map.value())` here?

CalebDu replied:
I plan to replace all get_ptr<T>() with tensor.data_ptr<T>()

bnellnm commented on csrc/moe/permute_unpermute_kernels/moe_permute_unpermute_kernel.h:
```diff
@@ -0,0 +1,508 @@
+  int64_t const stride = blockDim.x;
+  int64_t const num_elems_in_col = orig_cols / FINALIZE_ELEM_PER_THREAD;
+
+  // using BiasElem = cutlass::Array<ScaleBiasType, FINALIZE_ELEM_PER_THREAD>;
```
delete unused code?

bnellnm commented on csrc/moe/permute_unpermute_kernels/moe_permute_unpermute_kernel.h:
```diff
@@ -0,0 +1,508 @@
+  using InputElem = cutlass::Array<T, FINALIZE_ELEM_PER_THREAD>;
+  using OutputElem = cutlass::Array<OutputType, FINALIZE_ELEM_PER_THREAD>;
+  using ComputeElem = cutlass::Array<float, FINALIZE_ELEM_PER_THREAD>;
+  // auto const* bias_v = reinterpret_cast<BiasElem const*>(bias);
```
ditto

bnellnm commented on csrc/moe/permute_unpermute_kernels/moe_permute_unpermute_kernel.h:
```diff
@@ -0,0 +1,508 @@
+      ComputeElem expert_result = arrayConvert<InputElem, ComputeElem>(
+          expanded_permuted_rows_row_ptr[elem_index]);
+      thread_output = thread_output + row_scale * (expert_result);
+      has_valid = true;
```
Is this flag used by anything?

CalebDu replied:
This flag is unused. I'll delete it.

bnellnm commented on tests/kernels/test_moe_permute_unpermute.py:
```diff
@@ -0,0 +1,224 @@
+) -> list[torch.Tensor]:
+    # print(topk_ids)
+    n_token, n_hidden  = hidden_states.shape[0], hidden_states.shape[1]
+    ## torch sort is unstable
+    # sorted_topk_ids, sorted_indices = torch.sort(topk_ids.flatten()) 
```
there's a `stable=True` option on `torch.sort`

bnellnm commented on tests/kernels/test_moe_permute_unpermute.py:
```diff
@@ -0,0 +1,224 @@
+    align_block_size: Optional[int]=None,
+    fill_invalid_expert:int=-1
+) -> list[torch.Tensor]:
+    # print(topk_ids)
```
nit: remove debugging prints

bnellnm commented on tests/kernels/test_moe_permute_unpermute.py:
```diff
@@ -0,0 +1,224 @@
+                            rtol=0)
+
+
+# test_moe_permute_unpermute(4, 2048, 2, 4, 2, torch.float16, 128)
```
delete unused code.

CalebDu commented:
> @CalebDu overall the PR looks good. I mostly had minor comments. I was wondering if there were any size restrictions on the new kernels, e.g. do M (num tokens) or K (hidden dim) need to be multiples of 128, etc? If so, can you add some asserts to check any size restrictions?

I will follow your comments to update code.  `permute/unpermute` kernel only restrict K(hidden dim) * element_size(fp16/bf16/fp8) aligned to 16Byte because of cuda ld/st.128 vectorization instruction.

CalebDu committed:
update code according bnellnm's comment:
1. remove unused code
2. move all the non-trivial definitions from moe_permute_unpermute_kernel.h to .cu and .inl
3. some minor update

Signed-off-by: Caleb_Du <Caleb_Du@zju.edu.cn>

CalebDu commented:
@bnellnm I update code with your review.  But I do not get your idea about CUDA_Arch in cmakeLists.txt. Can you give more detail?

bnellnm commented:
> @bnellnm I update code with your review. But I do not get your idea about CUDA_Arch in cmakeLists.txt. Can you give more detail?

I meant using the preprocessor to avoid compiling/registering the new ops when they aren't supported (like you mentioned above doing something like `#if __CUDA_ARCH__ > 800`).  I think that would be more robust than checking the version strings inside cmake.

CalebDu committed:
remove arch limit in cmake and add return `token_expert_indices` in all invoking fused_topk code

Signed-off-by: Caleb_Du <Caleb_Du@zju.edu.cn>

CalebDu commented:
@bnellnm I remove arch limit in cmake. I review `permute/unpermute` kernel implementation. There is no arch speicialzed instruction like cp.async/tma/tensor core, only with cub.sort, ld/st and scalar compute with cuda core. So I think it's compatible for all cuda target and does not need preprocessor to isolate different cuda target.

bnellnm reviewed (commented):
There seems to be a few lint errors that need to be fixed but looks good otherwise!

CalebDu committed:
fix pre-commit  failed

Signed-off-by: Caleb_Du <Caleb_Du@zju.edu.cn>

tlrmchlsmth commented on csrc/moe/moe_permute_unpermute_op.cu:
```diff
@@ -0,0 +1,136 @@
+  TORCH_CHECK(
+      permuted_hidden_states.scalar_type() == hidden_states.scalar_type(),
+      "topk_ids dtype must be same as src_row_id2dst_row_id_map");
+  // TORCH_CHECK(permuted_hidden_states.size(0) == hidden_states.size(0) * topk,
+  //             "permuted_hidden_states must be [n_token * topk, n_hidden],"
+  //             "hidden_states must be [n_token, n_hidden]");
```
Debug cruft? Could you delete if so?

tlrmchlsmth commented on csrc/moe/permute_unpermute_kernels/moe_permute_unpermute_kernel.cu:
```diff
@@ -0,0 +1,229 @@
+                       int num_experts, int num_experts_per_node, int k,
+                       CubKeyValueSorter& sorter, void* sorter_ws,
+                       cudaStream_t stream) {
+  int64_t const expanded_num_rows = k * num_rows;
```
I think this should be:
```
  int64_t const expanded_num_rows = static_cast<int64_t>(k) * num_rows;
```
I think the way it is now, it won't cast until after the multiplication, which would allow for overflowing an int32

tlrmchlsmth reviewed (commented):
The implementation looks nice and clean. Are there any benchmark results?

(I did spot one potential overflow that should be addressed)

CalebDu committed:
1. fix call FusedMoE.select_experts failed
2. fix potential overflow and remove debug cruft with  tlrmchlsmth's review
3. add benchmark for performance

Signed-off-by: Caleb_Du <Caleb_Du@zju.edu.cn>

CalebDu commented on CMakeLists.txt:
```diff
@@ -661,6 +660,17 @@ if(VLLM_GPU_LANG STREQUAL "CUDA")
+      "csrc/moe/moe_permute_unpermute_op.cu")
+
+  set_gencode_flags_for_srcs(
+    SRCS "${MARLIN_PERMUTE_SRC}"
```
this line should be removed as bnell‘s review. I'll remove it.

CalebDu commented:
@tlrmchlsmth I update code with your review. And fix ci failed in calling FusedMoE.select_experts.
I add benchmark for comparison permute/unpermte customized kernel in **deepseek v3**  with  naive python implementation(`_moe_permute` `_moe_unpermute_and_reduce`) in deep_gemm_moe.py.
```
python vllm/benchmarks/kernels/benchmark_moe_permute_unpermute.py --dtype=fp8_w8a8  --model deepseek-ai/DeepSeek-V3 --trust-remote-code --use-customized-permute
```
# benchmark in H20 
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/caleb/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/caleb/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">

<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:12.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"Aptos Narrow", sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
-->

</head>

<body link="#467886" vlink="#96607D">


batch size: | permute customized | permute python | unpermute customized | unpermute python
-- | -- | -- | -- | --
Batch size: 1 | Permute time: 65.64 us | Permute time: 1449.89 us | Unpermute time: 11.29 us | Unpermute time: 13.10 us
Batch size: 2 | Permute time: 68.92 us | Permute time: 1449.52 us | Unpermute time: 11.45 us | Unpermute time: 12.98 us
Batch size: 4 | Permute time: 68.61 us | Permute time: 1449.67 us | Unpermute time: 11.71 us | Unpermute time: 13.48 us
Batch size: 8 | Permute time: 70.00 us | Permute time: 1451.41 us | Unpermute time: 12.54 us | Unpermute time: 15.40 us
Batch size: 16 | Permute time: 72.08 us | Permute time: 1451.32 us | Unpermute time: 12.34 us | Unpermute time: 21.45 us
Batch size: 24 | Permute time: 76.10 us | Permute time: 1456.61 us | Unpermute time: 12.37 us | Unpermute time: 27.91 us
Batch size: 32 | Permute time: 81.13 us | Permute time: 1456.31 us | Unpermute time: 12.52 us | Unpermute time: 33.58 us
Batch size: 48 | Permute time: 85.94 us | Permute time: 1465.27 us | Unpermute time: 12.66 us | Unpermute time: 45.62 us
Batch size: 64 | Permute time: 95.78 us | Permute time: 1469.02 us | Unpermute time: 12.77 us | Unpermute time: 60.06 us
Batch size: 96 | Permute time: 109.39 us | Permute time: 1479.88 us | Unpermute time: 14.10 us | Unpermute time: 84.60 us
Batch size: 128 | Permute time: 126.87 us | Permute time: 1491.65 us | Unpermute time: 14.32 us | Unpermute time: 110.02 us
Batch size: 256 | Permute time: 196.66 us | Permute time: 1537.67 us | Unpermute time: 17.87 us | Unpermute time: 226.25 us
Batch size: 512 | Permute time: 335.68 us | Permute time: 1628.52 us | Unpermute time: 28.91 us | Unpermute time: 455.32 us
Batch size: 1024 | Permute time: 629.52 us | Permute time: 1810.61 us | Unpermute time: 45.61 us | Unpermute time: 895.50 us
Batch size: 1536 | Permute time: 914.56 us | Permute time: 1997.25 us | Unpermute time: 52.90 us | Unpermute time: 1335.12 us
Batch size: 2048 | Permute time: 1198.34 us | Permute time: 2191.29 us | Unpermute time: 54.18 us | Unpermute time: 1777.86 us
Batch size: 3072 | Permute time: 1757.11 us | Permute time: 2583.16 us | Unpermute time: 63.88 us | Unpermute time: 2656.25 us
Batch size: 4096 | Permute time: 2307.67 us | Permute time: 3000.42 us | Unpermute time: 83.67 us | Unpermute time: 3535.37 us



</body>

</html>

tlrmchlsmth reviewed (approved):
LGTM, thanks for posting the performance numbers!
